File: examples/typescript/x402/clients/mcp/simple.ts
// Usage: pnpm tsx x402/clients/mcp/simple.ts

/**
 * Call x402-paid MCP tools with a CDP-managed wallet.
 *
 * `wrapMCPClientWithPayment` takes any `@modelcontextprotocol/sdk` Client and
 * an x402 payment client. Because `CdpX402Client` extends the base `x402Client`,
 * it drops in directly — CDP provisions the wallet and signs payments; no
 * private keys, no manual scheme registration.
 *
 * This minimal example connects, lists tools, then calls the free `ping` tool
 * and the paid `generate_report` tool. See `chatbot.ts` for an LLM-driven variant.
 *
 * Setup:
 *   1. Start the MCP server:  cd ../../servers/mcp && pnpm install && pnpm start
 *   2. Set CDP_API_KEY_ID, CDP_API_KEY_SECRET, CDP_WALLET_SECRET in your .env.
 *   3. Fund the printed EVM address with USDC on Base Sepolia (see the README).
 */
import "dotenv/config";

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { SSEClientTransport } from "@modelcontextprotocol/sdk/client/sse.js";
import { CdpX402Client } from "@coinbase/cdp-sdk/x402";
import { wrapMCPClientWithPayment } from "@x402/mcp";

const SERVER_URL = process.env.MCP_SERVER_URL ?? "http://localhost:4022";

async function main() {
  // A CDP-managed payment client — provisions the wallet lazily and signs 402s.
  // "development" registers Base Sepolia, since the MCP server settles there.
  const payment = new CdpX402Client({ environment: "development" });
  const { evmAddress } = await payment.getAddresses();
  console.log("CDP-managed x402 MCP client ready");
  console.log("  Paying from EVM address:", evmAddress);
  console.log("  Fund it with USDC on Base Sepolia before calling paid tools.\n");

  // Wrap a standard MCP client. Only callTool() is enhanced with payment;
  // every other method is a passthrough, so full MCP compatibility is kept.
  const mcp = wrapMCPClientWithPayment(
    new Client({ name: "cdp-x402-mcp-client", version: "1.0.0" }, { capabilities: {} }),
    payment,
    {
      autoPayment: true,
      onPaymentRequested: async ({ toolName, paymentRequired }) => {
        const price = paymentRequired.accepts[0];
        console.log(`Payment required for "${toolName}": ${price.amount} ${price.asset}`);
        console.log("  Approving...\n");
        return true; // return false to decline
      },
    },
  );

  await mcp.connect(new SSEClientTransport(new URL(`${SERVER_URL}/sse`)));
  console.log("Connected to MCP server at", SERVER_URL);

  const { tools } = await mcp.listTools();
  console.log("Available tools:", tools.map(t => t.name).join(", "), "\n");

  // Free tool — no payment.
  const ping = await mcp.callTool("ping");
  console.log("ping ->", ping.content[0]?.text, `(paymentMade: ${ping.paymentMade})\n`);

  // Paid tool — CdpX402Client signs and settles automatically.
  const report = await mcp.callTool("generate_report", { topic: "USDC on Base" });
  console.log("generate_report ->", report.content[0]?.text);
  if (report.paymentResponse?.transaction) {
    console.log("Payment settled, tx:", report.paymentResponse.transaction);
  }

  await mcp.close();
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});

File: examples/python/x402/clients/mcp/simple.py
# Usage: uv run python x402/clients/mcp/simple.py

"""Call x402-paid MCP tools with a CDP-managed wallet.

The signer is a CDP Server Wallet reached through eth_account's account
interface via EvmLocalAccount, registered onto a standard x402Client -- no
private keys. create_x402_mcp_client runs the 402 -> pay -> retry loop for you.

Setup:
  1. Start the server:  uv run python x402/servers/mcp/server.py
  2. Set CDP_API_KEY_ID, CDP_API_KEY_SECRET, CDP_WALLET_SECRET in examples/python/.env.
  3. Fund the printed address with USDC on Base Sepolia, or set
     CDP_FUND_FROM_FAUCET=true to self-fund on first run.

Run:
  uv run python x402/clients/mcp/simple.py
"""

import asyncio
import os

from cdp import CdpClient
from cdp.evm_local_account import EvmLocalAccount
from dotenv import load_dotenv
from x402 import x402Client
from x402.mcp import create_x402_mcp_client
from x402.mechanisms.evm import EthAccountSigner
from x402.mechanisms.evm.exact import ExactEvmScheme

load_dotenv()

SERVER_URL = os.getenv("MCP_SERVER_URL", "http://localhost:4022")
NETWORK = "eip155:84532"  # Base Sepolia


async def main() -> None:
    async with CdpClient() as cdp:
        account = await cdp.evm.get_or_create_account(name="x402-client-wallet-1")
        # EthAccountSigner adapts the CDP account to the x402 signer protocol.
        signer = EthAccountSigner(EvmLocalAccount(account))
        print(f"Paying from {signer.address}")

        # The same CDP credentials power both the faucet and the facilitator.
        if os.getenv("CDP_FUND_FROM_FAUCET", "").lower() == "true":
            tx = await cdp.evm.request_faucet(
                address=signer.address, network="base-sepolia", token="usdc"
            )
            print(f"Requested USDC from the CDP faucet: {tx}")
            print("Wait for it to confirm, then re-run without the flag to pay.")
            return

        payment_client = x402Client()
        payment_client.register(NETWORK, ExactEvmScheme(signer))

        async with create_x402_mcp_client(payment_client, SERVER_URL) as mcp:
            tools = (await mcp.list_tools()).tools
            print("Tools:", ", ".join(t.name for t in tools))

            ping = await mcp.call_tool("ping", {})
            print(f"ping -> {ping.content[0].text}")

            report = await mcp.call_tool("generate_report", {"topic": "USDC on Base"})
            print(f"generate_report -> {report.content[0].text}")
            # payment_response is a SettleResponse, or the raw dict if it could
            # not be parsed, so read the receipt without assuming either shape.
            receipt = report.payment_response
            tx_hash = (
                receipt.get("transaction")
                if isinstance(receipt, dict)
                else getattr(receipt, "transaction", None)
            )
            if tx_hash:
                print(f"Settled tx: {tx_hash}")


if __name__ == "__main__":
    asyncio.run(main())
